### ISSUE TITLE
[Bug]: "The folder contents could not be displayed" "Error" "/run/user/1000/doc/". Related to "xdg-document-portal".

### ISSUE THREAD

**Issue by Francewhoa** [flatpak/xdg-desktop-portal#1410]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.14.4

### What Linux distribution are you using?

Debian

### Linux distribution version

12 Bookworm. 6.1.0-18-amd64. Debian 6.1.76-1 (2024-02-01) x86_64 GNU/Linux.

### What architecture are you using?

x86_64

### How to reproduce

1. After a fresh reboot, use either **[LibreWolf](https://flathub.org/apps/io.gitlab.librewolf-community) 124.0.1-1** or **[Firefox](https://flathub.org/apps/org.mozilla.firefox) 124.0.2** to print to a PDF file any web page. This works smooth and fine for now.

1. Continue using the same device and apps above for multiple days without reboot. Do daily Flatpak updates without reboot. 

1. Until when you use either **LibreWolf 124.0.1-1** or **Firefox 124.0.2**, when you print the same web page to a PDF. 
  1. This error message is displayed: 
      1.  `The folder contents could not be displayed` `Error`[…] `/run/user/1000/doc/` […]
         1. This screenshot below shows this error in French. With steps to reproduce.
         
          ![Screenshot-38---cropped](https://github.com/flatpak/flatpak/assets/119942/5535f293-7174-40ae-8423-ecbfd8866d20)

      1. This error message above is the first challenge. Then the PDF file creation failed. This is the second challenge.
     
The same challenge can be reproduced the other way around with an incoming file instead of an outgoing file. For example, but not limited to, using the same Flatpak apps, uploading a PNG file to any webpages. The upload failed without an error message displayed. The same workaround below temporarily resolved this challenge.

Using:

* xdg-desktop-portal 1.16.0-2 from https://packages.debian.org/bookworm/xdg-desktop-portal
  * The above package includes the process **xdg-document-portal**. Which is the "desktop integration portal for Flatpak and Snap"
* Debian 12 Bookworm
* GNOME 43.9
* Wayland

### Expected Behavior

1. Error message is not displayed. 

1. File is created. Such as, but not limited to PDF file.

### Actual Behavior

1. The error message is displayed

1. Then the PDF file creation failed

### Additional Information

----------------

#### Temporary Workaround

For those facing this challenge, the steps below temporary fixed this challenge. But the same challenge comes back after a few hours or days.

Steps:

1. Using **GNOME System Monitor** or equivalent, kill this process `xdg-document-portal`

1. Fully exit the Flatpak app. Including all its windows.

1. Start the Flatpak app

1. Error is gone

----------------

#### Related Ticket

>bwrap: Can't find source path /run/user/1000/doc/by-app/com.spotify.Client: Permission denied flatpak/flatpak#5580

**Comment by ravilock** [comment]
I'm currently having the same (or a similar) issue with Fedora 41.
By digging on folders on `/run/user/1000/doc` and runing `ls -ld /run/user/1000/doc` i got:

`dr-x------ 2 my-user my-user 0 Dec 31  1969 /run/user/1000/doc`

I'm guessing that this is a bug considering that [this line of code](https://github.com/flatpak/xdg-desktop-portal/blob/main/document-portal/document-portal-fuse.c#L3582) seems to create the directory with `700` permissions.

Whenever a program installed via flatpak tries to use that mountpoint to write some files, I can see the following errors on `journalctl`

```
Jan 07 01:50:00 fedora PanGPUI.desktop[3836]: fusermount3: failed to access mountpoint /run/user/1000/doc: Permission denied
Jan 07 01:50:00 fedora PanGPUI.desktop[3812]: error: fuse init failed: Can't mount path /run/user/1000/doc
```

**Comment by Francewhoa** [comment]
Hello @swick 🙂 Thanks for both your reply and all your contribution to XDG-Destop-Portal. My reply is below.

---------

>Am I reading the comments here correctly that everyone affected is somehow starting two instances of x-d-p for the same user?

If by `x-d-p` you mean the process/instance `xdg-destop-portal`, then in my case, after a fresh reboot, there is only one instance `xdg-destop-portal`. After using Debian 12 with GNOME and Flatpak for a few hours, there is still only one process (instance) of `xdg-destop-portal`. After using the temporary workaround to restart `xdg-destop-portal`, there is still only one instance of `xdg-destop-portal`. I never recall ever manually starting a second `xdg-destop-portal` instance.

Related to this above, I also see `xdg-destop-portal-gtk`. Is `xdg-destop-portal-gtk` considered a second instance of `xdg-destop-portal`? My understanding is that `xdg-destop-portal-gtk` is just a service for `xdg-destop-portal`. 

For those not familiar with `xdg-destop-portal-gtk`, [GTK](https://www.gtk.org/) means GNOME. Details about the difference between  `xdg-destop-portal-gtk`and `xdg-destop-portal` at:

*  `xdg-destop-portal-gtk` package at https://packages.debian.org/bookworm/xdg-desktop-portal-gtk
*  `xdg-destop-portal` package at https://packages.debian.org/bookworm/xdg-desktop-portal

**Comment by swick** [comment]
Is this only happening after a new instance of xdg-document-portal gets launched somehow? I.e. do you all have something in your logs about xdg-document-portal getting killed and restarted?

**Comment by swick** [comment]
Understood. This happens without any component restarting. The problem is that I can't reproduce the issue, so I have to rely on information from the people who can.

To help with debugging, please use `systemctl edit --user xdg-desktop-portal.service` and `systemctl edit --user xdg-document-portal.service` and add:

```
[Service]
Environment="G_MESSAGES_DEBUG=all"
```

Then restart the system.

If you then run into this issue again please:

* Do not restart, kill, etc. anything before the steps below
* Get the logs of xdg-desktop-portal and xdg-document-portal from the journal (`journalctl --user -I --unit xdg-desktop-portal.service`)

**Comment by Francewhoa** [comment]
@swick, I do not know what is the cause of this challenge #1410. Which happens [before](https://github.com/flatpak/xdg-desktop-portal/issues/1410#steps-to-reproduce) end-users manually restart the instance `xdg-document-portal` (XDP). Could it be caused by either Flatpak engine updates or Flatpak app updates only for already opened Flatpak apps?

I am asking this above because, what we know so far:

  * This challenge #1410 happens **[before](https://github.com/flatpak/xdg-desktop-portal/issues/1410#steps-to-reproduce)** end-users manually restart the process/instance `xdg-document-portal` (XDP). 
      
  * This challenge also happens **[after](https://github.com/flatpak/xdg-desktop-portal/issues/1410#steps-to-reproduce-after-manual-restart)** end-users manually restart XDP.
      
Could this challenge be caused by Flatpak update which somehow automatically restart XDP? In turn, already opened apps looses their permission to save to `/doc/<RAMDOM OLD ID>/` folder? In other words, those steps in chronological order:

  1. Fresh reboot
         
  1. Using Debian 12 Bookworm, using Wayland, using LibreWolf, print to PDF. This is successful. No error. Leave LibreWolf open.
         
  1. Wait a few days, until a new LibreWolf version is available. Then update it. During this Flatpak update, the theory is that, somehow, the Flatpak engine automatically restart XDP. During this update, Flatpak creates a new authorized `/doc/<RAMDOM NEW ID>/` folder. Denies access to the `/doc/<RAMDOM OLD ID>/` folder. In turn, the already open Flatpak apps, are no longer able to save to their old authorized `/doc/<RAMDOM OLD ID>/` folder. Why? Because that authorized path has changed. In other words, before the Flatpak update, the already opened Flatpak apps were authorized to access the `/doc/<RAMDOM OLD ID>/`. But after the Flatpak update, those same already opened apps are not authorized to save to the `/doc/<RAMDOM OLD ID>/`.
         
  1. If this theory is correct, to resolve this challenge, during Flatpak update, how about securely remember the apps authorized to accesses `/doc/<RAMDOM OLD ID>/` and continue those already authorized accesses to `/doc/<RAMDOM OLD ID>/`? If this is not appropriate or not secured, during a flatpak update, how about at minimum, somehow detect the presently open flatpak apps and end-users that, for those already apps to remain stable, they need to be restarted?

**Comment by swick** [comment]
> Wait a few days, until a new LibreWolf version is available or a new Flatpak engine version is available. Then update it. During this Flatpak update, the theory is that, somehow, the Flatpak engine automatically restart XDP instance

Yeah, I think you're onto something. If the package manager decides to restart xdg-document-portal because it, or a dependency has been updated, the existing mount that is mounted into the sandbox is going away, and the new mount can't get mounted into existing sandboxes.

This explains it, but unfortunately there isn't much we can do. Live updating a system comes with inherent issues; I suggest to only update when (soft)rebooting.
